### PR TITLE
align behavior of constant expressions in initializers of ambient enu…

### LINKED ISSUE #2790: Const enum expression in Ambient enum declaration behaves differently between spec and compiler
According to spec:

> An expression is considered a constant enum expression if it is one of the following:
> ....
> *** An identifier or property access that denotes a previously declared member in the same constant enum declaration.
> ....
 
In the Ambient Enum Declaration,

> In ambient enum declarations, all values specified in enum member declarations must be classified as constant enum expressions.

However, compiler reports an error for the following code:
```ts
declare enum E {
    a = 10,
    b = a,  // Ambient enum elements can only have integer literal initializers.
    e = 10 << 2 * 8,  // error as well
}
```

This is contradict to what it means to be constant enum expression

### PR DESCRIPTION
**Pull request by vladima**
…m members with spec. Fixes #2790

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -12940,26 +12940,41 @@ namespace ts {
...
                     let initializer = member.initializer;
                     if (initializer) {
                         autoValue = computeConstantValueForEnumMemberInitializer(initializer, enumType, enumIsConst, ambient);
                     }
                     else if (ambient && !enumIsConst) {
+                        // In ambient enum declarations that specify no const modifier, enum member declarations 
+                        // that omit a value are considered computed members (as opposed to having auto-incremented values assigned).
                         autoValue = undefined;
                     }
+                    else if (previousEnumMemberIsNonConstant) {
```
Comment (yuit) [inline src/compiler/checker.ts]: nit: since u only use this variable here, just do ```autoValue === undefined```

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -12940,26 +12940,41 @@ namespace ts {
             if (!(nodeLinks.flags & NodeCheckFlags.EnumValuesComputed)) {
                 let enumSymbol = getSymbolOfNode(node);
                 let enumType = getDeclaredTypeOfSymbol(enumSymbol);
-                let autoValue = 0;
+                let autoValue = 0; // set to undefined when enum member is non-constant
```
Comment (yuit) [inline src/compiler/checker.ts]: comment :+1:

File: src/compiler/checker.ts (hunk 1 above)
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: I'd prefer to have the constant since it makes it clearer

File: src/compiler/diagnosticMessages.json [hunk 3]
```diff
@@ -195,7 +195,7 @@
         "category": "Error",
         "code": 1063
     },
-    "Ambient enum elements can only have integer literal initializers.": {
+    "In ambient enum declarations member initializer must be constant expression.": {
```
Comment (yuit) [inline src/compiler/diagnosticMessages.json]: ```In ambient enum declarations, member initializer must be constant enum expression``` Use similar vocabulary as spec

**Comment by yuit** [comment]
:+1:

**Comment by vladima** [comment]
@DanielRosenwasser any other comments?

File: src/compiler/diagnosticMessages.json (hunk 3 above)
Comment (DanielRosenwasser) [inline src/compiler/diagnosticMessages.json]: An ambient enum member must be initialized with a constant enum expression.

Comment (vladima) [inline src/compiler/diagnosticMessages.json]: Current message is based on the text that we show for const enums:
    "In 'const' enum declarations member initializer must be constant expression.": {

**Comment by DanielRosenwasser** [comment]
:+1:

File: src/compiler/diagnosticMessages.json (hunk 3 above)
Comment (mhegazy) [inline src/compiler/diagnosticMessages.json]: nit. "Enum member initializer must be constant expression in an ambient enum declaration"

the emphasis should be on the enum member initializer and not the ambient enum declaration.

**Pull request merged**